core/src/main/java/org/apache/brooklyn/core/catalog/internal/BasicBrooklynCatalog.java

```diff
@@ -558,15 +561,34 @@ private void collectCatalogItems(String sourceYaml, Map<?,?> itemMetadata, List<
             throw new IllegalStateException("Could not resolve plan once id and itemType are known (recursive reference?): "+sourceYaml);
         }
         String sourcePlanYaml = planInterpreter.getPlanYaml();
-        
-        CatalogItemDtoAbstract<?, ?> dto = createItemBuilder(itemType, symbolicName, version)
+
+        CatalogItemBuilder itemBuilder = createItemBuilder(itemType, symbolicName, version)
             .libraries(libraryBundles)
             .displayName(displayName)
             .description(description)
             .deprecated(catalogDeprecated)
             .iconUrl(catalogIconUrl)
-            .plan(sourcePlanYaml)
-            .build();
+            .plan(sourcePlanYaml);
+        if (itemType.equals(CatalogItemType.ENTITY) || itemType.equals(CatalogItemType.TEMPLATE)) {
+            String stringServiceType = getFirstAs(itemAsMap, String.class, "type").orNull();
+            // This is not a single item but a services based YAML
+            if (stringServiceType == null) {
+                final List<Map<?, ?>> services = getFirstAs(itemAsMap, List.class, "services").orNull();
+                if (services.isEmpty()) {
+                    throw new IllegalStateException("Services are empty: " + sourceYaml);
+                }
+                stringServiceType = services.get(0).get("type").toString();
+            }
+            Class<?> serviceType = JavaBrooklynClassLoadingContext.create(mgmt).tryLoadClass(stringServiceType).orNull();
+            if (serviceType == null) {
+                final CatalogItem<?, ?> relatedCatalogItem = getCatalogItem(stringServiceType, "");
+                if (relatedCatalogItem != null) {
+                    serviceType = JavaBrooklynClassLoadingContext.create(mgmt).tryLoadClass(relatedCatalogItem.getJavaType()).orNull();
```

Comment:
Not convinced by the use of `JavaBrooklynClassLoadingContext`. What happens when the service is in a bundle? @neykov Is a suitable `OsgiBrooklynClassLoadingContext` available in this method (BasicBrooklynCatalog.collectCatalogItems)? Or perhaps you could check the catalogue item's bundles too.